From df0dbf72f1d3042ac98215b947376d571616a877 Mon Sep 17 00:00:00 2001
From: Arne Coucheron <arco68@gmail.com>
Date: Fri, 23 Oct 2015 02:04:56 +0200
Subject: [PATCH] Shim library for missing symbols in Samsung blobs

 * Provide missing symbol needed for Samsung RIL
 * rild binary and libril.so library needs to be replaced by ones
   from a lollipop ROM. The ones from kitkat will simply crash.

Change-Id: I50d4c89727950e1a8c238383e77e3855333e43f9
---
 libsamsung_symbols/Android.mk      | 28 ++++++++++++++++++++++++++++
 libsamsung_symbols/samsung_ril.cpp | 21 +++++++++++++++++++++
 rootdir/init.qcom.rc               |  2 ++
 serrano-common.mk                  |  4 ++++
 4 files changed, 55 insertions(+)
 create mode 100644 libsamsung_symbols/Android.mk
 create mode 100644 libsamsung_symbols/samsung_ril.cpp

diff --git a/libsamsung_symbols/Android.mk b/libsamsung_symbols/Android.mk
new file mode 100644
index 0000000..c2e942e
--- /dev/null
+++ b/libsamsung_symbols/Android.mk
@@ -0,0 +1,28 @@
+# Copyright (C) 2015 The CyanogenMod Project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+LOCAL_PATH := $(call my-dir)
+
+include $(CLEAR_VARS)
+
+LOCAL_SRC_FILES := \
+    samsung_ril.cpp
+
+LOCAL_SHARED_LIBRARIES := libbinder
+
+LOCAL_MODULE := libsamsung_symbols
+LOCAL_MODULE_TAGS := optional
+LOCAL_MODULE_CLASS := SHARED_LIBRARIES
+
+include $(BUILD_SHARED_LIBRARY)
diff --git a/libsamsung_symbols/samsung_ril.cpp b/libsamsung_symbols/samsung_ril.cpp
new file mode 100644
index 0000000..404616d
--- /dev/null
+++ b/libsamsung_symbols/samsung_ril.cpp
@@ -0,0 +1,21 @@
+/*
+ * Copyright (C) 2015 The CyanogenMod Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* status_t Parcel::writeString16 */
+extern "C" int _ZN7android6Parcel13writeString16EPKDsj();
+extern "C" int _ZN7android6Parcel13writeString16EPKtj() {
+    return _ZN7android6Parcel13writeString16EPKDsj();
+}
diff --git a/rootdir/init.qcom.rc b/rootdir/init.qcom.rc
index 727fc50..7c2971a 100644
--- a/rootdir/init.qcom.rc
+++ b/rootdir/init.qcom.rc
@@ -31,6 +31,8 @@ import init.target.rc
 import init.carrier.rc
 
 on init
+    export LD_PRELOAD "/system/lib/libsamsung_symbols.so"
+
     # Set permissions for persist partition
     mkdir /persist 0771 system system
 
diff --git a/serrano-common.mk b/serrano-common.mk
index 10d62ed..7cf43fb 100644
--- a/serrano-common.mk
+++ b/serrano-common.mk
@@ -113,6 +113,10 @@ PRODUCT_PACKAGES += \
     fsck.f2fs \
     mkfs.f2fs
 
+# Samsung symbols
+PRODUCT_PACKAGES += \
+    libsamsung_symbols
+
 # IR feature permission
 PRODUCT_COPY_FILES += \
     frameworks/native/data/etc/android.hardware.consumerir.xml:system/etc/permissions/android.hardware.consumerir.xml
